src/main/kotlin/com/sourcegraph/cody/edit/FixupSession.kt

```diff
@@ -0,0 +1,354 @@
+      agent.server.commandExecute(CommandExecuteParams(COMMAND_UNDO, listOf(taskId!!)))
+    }
+    undoEdits()
+    finish()
+  }
+
+  fun performInlineEdits(edits: List<TextEdit>) {
+    val project = editor.project ?: return
+    val doc: Document = editor.document
+
+    // TODO: This is an artifact of the update to concurrent editing tasks.
+    // We do need to mute any LensGroup listeners, but this is an ugly way to do it.
+    // There are multiple Lens groups; we need a Document-level listener list.
+    lensGroup?.withListenersMuted {
+      if (!controller.isEligibleForInlineEdit(editor)) {
+        return@withListenersMuted logger.warn("Inline edit not eligible")
+      }
+      // Mark all the edit locations so the markers will move as we edit the document,
+      // preserving the original positions of the edits.
+      val markers = edits.mapNotNull { createMarkerForEdit(doc, it) }
+      val sortedEdits = edits.zip(markers).sortedByDescending { it.second.startOffset }
+      // Apply the edits in a write action.
+      WriteCommandAction.runWriteCommandAction(project) {
+        for ((edit, marker) in sortedEdits) {
+          when (edit.type) {
+            "replace",
+            "delete" -> ReplaceUndoableAction(this, edit, marker)
+            "insert" -> InsertUndoableAction(this, edit, marker)
+            else -> logger.warn("Unknown edit type: ${edit.type}")
+          }
+        }
+      }
+    }
+  }
+
+  private fun createMarkerForEdit(doc: Document, edit: TextEdit): RangeMarker? {
+    val startOffset: Int
+    val endOffset: Int
+    when (edit.type) {
+      "replace",
+      "delete" -> {
+        val range = edit.range ?: return null
+        startOffset = doc.getLineStartOffset(range.start.line) + range.start.character
+        endOffset = doc.getLineStartOffset(range.end.line) + range.end.character
+      }
+      "insert" -> {
+        val position = edit.position ?: return null
+        startOffset = doc.getLineStartOffset(position.line) + position.character
+        endOffset = startOffset
+      }
+      else -> return null
+    }
+    return createMarker(startOffset, endOffset)
+  }
+
+  fun createMarker(startOffset: Int, endOffset: Int): RangeMarker {
+    return editor.document.createRangeMarker(startOffset, endOffset).apply {
+      rangeMarkers.add(this)
+    }
+  }
+
+  fun removeMarker(marker: RangeMarker) {
+    try {
+      marker.dispose()
+      rangeMarkers.remove(marker)
+    } catch (x: Exception) {
+      logger.debug("Error disposing marker $marker", x)
+    }
+  }
+
+  protected fun undoEdits() {
+    val project = editor.project ?: return
+    if (project.isDisposed) return
+    val fileEditor = getEditorForDocument(editor.document, project)
+    val undoManager = UndoManager.getInstance(project)
+    if (undoManager.isUndoAvailable(fileEditor)) {
+      undoManager.undo(fileEditor)
+    }
+  }
+
+  private fun getEditorForDocument(document: Document, project: Project): FileEditor? {
```

Comment:
A lot of method signatures like this one would simplify a lot if you will make `document: Document`, and  `project: Project` parameters of `FixupSession` constructor.

Reply:
Although I do use the editor in at least three other situations in this file, you are absolutely correct that passing in non-null project and document parameters to the constructor will significantly simplify the rest of the code in the file. So I went ahead and did that, and pushed a fix to this branch.